### PR TITLE
Support dependencies to other Feature Variants (including Test Fixtures)

### LINKED ISSUE #298: Plugin incorrectly(?) reporting unused dependency
**Build scan link**
https://scans.gradle.com/s/wqljzkve3gazi

**Plugin version**
0.61.0

**Gradle version**
gradle-6.8-branch-eskatos_kotlin_embedded_1_4-20200922125722+0000

**Android Gradle Plugin (AGP) version**
N/A

**Describe the bug**
Getting a failure claiming I have an unused dependency that I think is invalid - possibly due to the plugin not knowing about the `testFixtures` configuration? Or not knowing about how kotlin works with top level functions?

**To Reproduce**
1. Check out git@github.com:Mahoney-playground/goos.git
2. Run ./gradlew build

**Expected behavior**
I think the build should pass

**Additional context**
https://docs.gradle.org/current/userguide/java_testing.html#sec:java_test_fixtures

### LINKED ISSUE #836: Incorrect advice when using a dependency with a capability
**Build scan link**
Private :/

**Plugin version**
`1.17.1-SNAPSHOT`

**Gradle version**
`7.4.2`

**(Optional) Android Gradle Plugin (AGP) version**
Not applicable.

**`reason` output for bugs relating to incorrect advice**

```
dependencies {
    implementation(project(":bar")) {
        capabilities {
            requireCapability("com.foo:bar-<capability>")
        }
    }
}
```

```
> Transitively used dependencies that should be declared directly as indicated:
    implementation project(':bar')
```

**Describe the bug**
When computing the advice in `StandardTransform#computeAdvice` the `usages` are compared to the `declarations` and based on what is incorrect some advice is given. However the set of `declarations` have filtering applied to `special dependencies like test fixtures or platforms`, but this filtering does not happen for the `usages`. This result in incorrect advice where a dependency on a 'special dependency' is used and declared, the declaration is filtered out when computing the advice and then the plugin advises to declare the dependency. Unfortunately the advice prints the advice also incorrectly as the capability of the dependency is not taken into account.

**To Reproduce**
https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/commit/35ebb6bb20d1aab8e8bbae6eacec9ea4f57e5b0d

**Expected behavior**
Either
* Ignore the usage of the 'special' dependency with a capability
* Include the capability in the dependency
```
> Transitively used dependencies that should be declared directly as indicated:
    implementation(project(':bar')) { capabilities { requireCapability("com.foo:bar-<capability>") }
```

**Additional context**
This relates to #298, but it's slightly different as 'special' dependencies are not supported and should be filtered out.

I've managed to trace this back to `com.autonomousapps.internal.utils.GradleStringsKt#toCoordinates(org.gradle.api.artifacts.result.ResolvedDependencyResult)` where the `ResolvedDependencyResult` is converted to the internal model `Coordinates`. The `ResolvedDependencyResult` might also have `requested.requestedCapabilities`, but this information is not saved to `Coordinates`. Something like the following could be added to determine whether a capability needs to be saved to the `Coordinates`. 
```
val requestedCapability = requested.requestedCapabilities.firstOrNull()
  val selectedCapability = selected.variants.firstOrNull()?.capabilities?.firstOrNull()
  if (requestedCapability == selectedCapability) {
    capability = selectedCapability?.name
  }
```

### LINKED ISSUE #841: False positive for testImplementation dependency with classifier
**Build scan link**
N/A

**Plugin version**
1.18.0

**Gradle version**
7.6

**(Optional) Android Gradle Plugin (AGP) version**
N/A

**`reason` output for bugs relating to incorrect advice**
```
> Task :E84HSP-app:reason

----------------------------------------
You asked about the dependency 'se.micronic.bennu:kernel:6.2.0.0-46-16c1002'.
You have been advised to remove this dependency from 'testImplementation'.
----------------------------------------

Shortest path from :E84HSP-app to se.micronic.bennu:kernel:6.2.0.0-46-16c1002 for compileClasspath:
:E84HSP-app
\--- se.micronic.bennu:kernel:6.2.0.0-46-16c1002

Shortest path from :E84HSP-app to se.micronic.bennu:kernel:6.2.0.0-46-16c1002 for runtimeClasspath:
:E84HSP-app
\--- se.micronic.bennu:kernel:6.2.0.0-46-16c1002

Shortest path from :E84HSP-app to se.micronic.bennu:kernel:6.2.0.0-46-16c1002 for testCompileClasspath:
:E84HSP-app
\--- se.micronic.bennu:kernel:6.2.0.0-46-16c1002

Shortest path from :E84HSP-app to se.micronic.bennu:kernel:6.2.0.0-46-16c1002 for testRuntimeClasspath:
:E84HSP-app
\--- se.micronic.bennu:kernel:6.2.0.0-46-16c1002

Source: main
------------
* Exposes 27 classes, 5 of which are shown: se.micronic.bennu.kernel.gui.DisplayConditionsIfc, se.micronic.bennu.kernel.logevent.LogEventDeclaration, se.micronic.bennu.kernel.logevent.LogEventId, se.micronic.bennu.kernel.logevent.XMLLogEventSpecificationBase, se.micronic.bennu.kernel.message.Command (implies api).
* Imports 3 constants: se.micronic.bennu.kernel.message.*, se.micronic.bennu.kernel.message.CommandDeclaration.*, se.micronic.bennu.kernel.variable.* (implies implementation).

Source: test
\------------
\* Uses 1 class: se.micronic.bennu.kernel.message.threaded.ThreadedCommandTestExtension (implies testImplementation).

```
**Describe the bug**
I have this dependency, (please note the classifier = 'test'), without it my project do not compile.
```
  testImplementation("se.micronic.bennu:kernel:6.2.0.0-46-16c1002") { artifact { classifier = 'test' } }
```

So, the test code of my project depends on the test code of another project which is selected using the 'classifier = test', however the dependency analysis seems to ignore the classifier.

**To Reproduce**
N/A

**Expected behavior**
Should not give advise to remove the dependency since it is used for compilation of test code.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by jjohannes**
The main change in this PR is that it adds something I called `GradleVariantIdentification` to all `Coordinates`.

- On the declaration side (consumer) The `GradleVariantIdentification` collects all [capabilities](https://docs.gradle.org/current/userguide/feature_variants.html#sec::consuming_feature_variants) and [attributes](https://docs.gradle.org/current/userguide/variant_attributes.html) that are defined on a dependency. Shortcut notations like `platform(...)` and `testFixtures(...)` also add these under the hood.
- On the resolved graph side (producer) the `GradleVariantIdentification` contains all the _capabilities_ of the selected node. The attributes are ignored here as we do not need them (see also comments in code).

With this, the analysis plugin now recognizes the fact that it is possible to have multiple nodes with the same `identifier` but different `GradleVariantIdentification` in the graph. This is the case when the capabilities differ. Because it is allowed to have multiple variants of one component in the graph as long as they have different capabilities. For example, there can be two `org.example:foo` nodes: one with capability `org.example:foo` and one with capability `org.example:foo-test-fixtures`.

The most critical impact on the implementation is IMO that there are several places throughout the code base that assume/assumed that Coordinates are comparable for different reasons. We need to be careful now, as different coordinates can mean the same. For example, if I don't declare any capability, the "main" capability is automatically used. The "Declaration Coordinates" then do not have a capability, but the "Resolved Graph Coordinates" will have the "main" capability. 

Therefore, several places needed to be touched, but I am confident wit the implementation now. I tried several things before I ended up with this solution (can still be seen in the commit history). What we have now feels most sound and also future-proof. We now collect all information we need to identify a component+variant and then can use this later during advice computation to make decisions. This should cover all "Feature Variant" standard cases. But if we need more information in the future, we can extend what we put into `GradleVariantIdentification`. (For example, it could also contain 'classifier' information – #342 – if we could somehow get this for 'both sides'.)

I chose the name `GradleVariantIdentification` as the _main_ thing, because _capability_ already has a different meaning in the context of this plugin. But I am happy to rename that if there are better ideas. 😄 

The PR resolves and adds test coverage for:
- Resolves #298 🎉 
- Fixes #836 
- Gives a alternative solution for #859 (will comment there)
- Fixes #841 in so far that there are no more false positives (we still won't be able to suggest classifiers correctly, but that is #342) will comment there

### CODE REVIEW

**Comment by autonomousapps** [comment]
I'd prefer to wait on reviewing this till the other PR gets merged and this rebased.

File: src/main/kotlin/com/autonomousapps/tasks/SynthesizeDependenciesTask.kt
Comment (jjohannes) [inline src/main/kotlin/com/autonomousapps/tasks/SynthesizeDependenciesTask.kt:217]: This + the `Capability.merge()` method is the fix for #841

File: src/main/kotlin/com/autonomousapps/model/Coordinates.kt
Comment (jjohannes) [inline src/main/kotlin/com/autonomousapps/model/Coordinates.kt:55]: With all the different solutions I considered, I always tried to reduce assumptions and guesswork on top of the information we get out of Gradle. This method is basically what is left and for which I don't know if and how it could be further improved.
This is what the user gets in case of an ADD advice where we do not have an existing declaration to "reuse". I think this is a good start and if users report cases with unexpected advices, this can be tuned further if we have real-world reproducers.

**Comment by jjohannes** [comment]
I did some more manual testing and there still seems to be an issue when the **same** dependency is declared **from** different source sets in the same project. I'll see that I provide tests that reproduce this (+ a fix) soon.

**Comment by jjohannes** [comment]
With this (actually with #852 already 😬) users might suddenly get more in their reports than before as the plugin is now doing more analysis as before by default. I wonder if there should at least be an opt-out. There were two things I was thinking about. (Which could be done later. But maybe one of them before this is released...)

- Should the `dependencyAnalysis.issues.only... {}` mechanism be extended to allow selective ignoring per source set? Right now, it works for all source sets, but you can not say for example "do everything for main, but ignore for testFixtures".
- Should the `dependency.analysis.test.analysis=false` flag be generalized to something that's configured in the `dependencyAnalysis` extension and where you with which you can ignore arbitrary source sets (not only test).

**Comment by TimvdLippe** [comment]
Commenting here as I originally asked Jendrik about this.

While trying to roll out the new changes, we were indeed hit by new warnings (which indeed is already true with #852). However, we weren't able to selectively ignore the warnings, other than ignoring all warnings. However, we don't want to lose coverage for `main` while we are cleaning up `test` (or any other sourceset). So any mechanism that allows for more selective filtering would be great. I have no strong preference on how that mechanism should work.

File: src/main/kotlin/com/autonomousapps/tasks/ComputeAdviceTask.kt
Comment (jjohannes) [inline src/main/kotlin/com/autonomousapps/tasks/ComputeAdviceTask.kt:261]: Not sure if there is a better place to but this exception. I think it's too in individual to solve this with 'bundles' (as it's done in other cases).
For reference: https://github.com/gradle/gradle/blob/d9303339298e6206182fd1f5c7e51f11e4bdff30/subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaTestFixturesPlugin.java#L68

Probably was a mistake to hardcode this into the `test-fixtures-plugin` ... 😐

**Comment by jjohannes** [comment]
Regarding the "opt-out". I prototyped this, that would turn off the whole analysis for a given _source set_. Either for _all_ or selected projects:

```
dependencyAnalysis {
    issues {
        all {
            ignoreSourceSet("mockApi")
            ignoreSourceSet("endToEndTest")
        }
    }
}
```

Can try it out from here: https://github.com/jjohannes/dependency-analysis-android-gradle-plugin/tree/ignore-selected-source-sets

If this is wanted, I can add a few tests and do follow up PR, once this one is integrated.

File: src/main/kotlin/com/autonomousapps/model/Coordinates.kt
Comment (jjohannes) [inline src/main/kotlin/com/autonomousapps/model/Coordinates.kt:9]: Should we call the field just `gvi` (or smth completely different) to make this more compact in the Json files?

**Comment by ljacomet** [comment]
> Regarding the "opt-out".

It would be great if the opt out for _all_ projects was lenient: If a source set is referenced but does not exist in a specific project, simply ignore.

File: src/main/kotlin/com/autonomousapps/tasks/SynthesizeProjectViewTask.kt
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/tasks/SynthesizeProjectViewTask.kt:195]: clever

File: src/test/kotlin/com/autonomousapps/transform/StandardTransformTest.kt
Comment (autonomousapps) [inline src/test/kotlin/com/autonomousapps/transform/StandardTransformTest.kt:1]: It doesn't look like you added any new tests here, is that correct?

**Review by autonomousapps — COMMENTED** [review]
Thanks for this! And sorry it took so long to review. Left a few minor comments. Just want to hear your response before I approve.

File: src/main/kotlin/com/autonomousapps/internal/advice/AdvicePrinter.kt
Comment (jjohannes) [reply src/main/kotlin/com/autonomousapps/internal/advice/AdvicePrinter.kt:14]: Then `when` is now done in `fun gav(coordinates: Coordinates)` already (see changes below).

This is because we might need to add a `{ ... }` configuration block as part of the "printableIdentifier" for a `requireCapability`. Because of that, we need to make the decision if we want to add `(` and `)` there (as `)` might be in the middle of the string before the `{`).

File: src/main/kotlin/com/autonomousapps/internal/advice/AdvicePrinter.kt
Comment (jjohannes) [reply src/main/kotlin/com/autonomousapps/internal/advice/AdvicePrinter.kt:18]: See https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/pull/854/files#r1158053173

File: src/main/kotlin/com/autonomousapps/model/Coordinates.kt
Comment (jjohannes) [reply src/main/kotlin/com/autonomousapps/model/Coordinates.kt:9]: I don't have a strong opinion either. I also don't have a concrete case where anyone "complained". I just thought that the size of the files were a concern in the past already. I remember you thinking about alternative formats (but maybe I remember that wrongly).

Let's leave it like this for now then.

File: src/main/kotlin/com/autonomousapps/tasks/SynthesizeProjectViewTask.kt
Comment (jjohannes) [reply src/main/kotlin/com/autonomousapps/tasks/SynthesizeProjectViewTask.kt:195]: I am actually not 100% sure why it needs to be like this on Android. But there were tests showing cases that needed this.

In "plain JVM" projects this cannot be ignored to support all cases that are shown in the tests this PR adds. So I think this distinction here is fine for now.

File: src/test/kotlin/com/autonomousapps/transform/StandardTransformTest.kt
Comment (jjohannes) [reply src/test/kotlin/com/autonomousapps/transform/StandardTransformTest.kt:1]: I did not add a test. I adjusted the exiting tests as they now also have to respect the distinction between "Declaration Coordinates" and "Resolved Graph Coordinates". See this part of the PR description:

> The most critical impact on the implementation is IMO that there are several places throughout the code base that assume/assumed that Coordinates are comparable for different reasons. We need to be careful now, as different coordinates can mean the same. For example, if I don't declare any capability, the "main" capability is automatically used. The "Declaration Coordinates" then do not have a capability, but the "Resolved Graph Coordinates" will have the "main" capability.

**Comment by jjohannes** [comment]
Thanks @autonomousapps! I responded to the comments. Let me know if there are any any concerns left before integrating this.

After this is merged, it would be great if you could also integrate the (small) follow-up in #882 before the next release.

**Review by autonomousapps — APPROVED** [review]
Thanks! Sorry for the approval delay. I was expecting a github notification that you had responded to my comments, but it never arrived 😕

**Pull request merged**
